File: leetcode/p0211/Solution.java
class WordDictionary {
    static class TrieNode {
        boolean isEnd;
        TrieNode[] child = new TrieNode[26];
    }
    
    private TrieNode root = new TrieNode();
    
    public WordDictionary() {
        
    }
    
    public void addWord(String word) {
        TrieNode cur = root;
        
        for (char c : word.toCharArray()) {
            int index = c - 'a';
            if (cur.child[index] == null) {
                cur.child[index] = new TrieNode();
            }
            cur = cur.child[index];
        }
        cur.isEnd = true;
    }
    
    public boolean search(String word) {
        return search(root, word);
    }
     
    public boolean search(TrieNode node, String word) {
        TrieNode cur = node;
        
        for (int i = 0; i < word.length(); ++i) {
            char c = word.charAt(i);
            if (c == '.') {
                for (int j = 0; j < cur.child.length; ++j) {
                    if (cur.child[j] != null && search(cur.child[j], word.substring(i + 1))) {
                        return true;
                    }
                }
                return false;
            } else if (cur.child[c - 'a'] == null) {
                return false;
            }
            cur = cur.child[c - 'a'];
        }
        return cur.isEnd;     
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */

File: leetcode/p0211/Solution.kt
/*
class WordDictionary() {
    private val set = mutableSetOf<String>()
    
    fun addWord(word: String) {
        set.add(word)
        
        val chars = word.toCharArray()
        for (i in 0 until chars.size) {
            val temp = chars[i]
            chars[i] = '.'
            set.add(String(chars))
            chars[i] = temp
        }
    }

    fun search(word: String): Boolean {
        return set.contains(word)
    }

}
*/

class TrieNode(var isEnd:Boolean = false, val child:Array<TrieNode?> = Array<TrieNode?>(26) {null})

class WordDictionary() {
    private val root = TrieNode()
    fun addWord(word: String) {
        var cur:TrieNode = root
        val chars = word.toCharArray()
        for (c in word.toCharArray()) {
            if (cur.child[c - 'a'] == null) {
                cur.child[c - 'a'] = TrieNode()
            }
            cur = cur.child[c - 'a']!!
        }
        cur.isEnd = true
    }

    fun search(word: String): Boolean {
        return search(root, word)
    }
    
    private fun search(node:TrieNode, word: String): Boolean {
        var cur:TrieNode = node
        for (i in 0 until word.length) {
            if (word[i] == '.') {
                for (j in 0 until cur.child.size) {
                    if (cur.child[j] != null && search(cur.child[j]!!, word.substring(i + 1))) {
                        return true
                    }
                }
                return false
            } else if (cur.child[word[i] - 'a'] == null) {
                return false
            }
            cur = cur.child[word[i] - 'a']!!
        }
        return cur.isEnd
    }

}


/**
 * Your WordDictionary object will be instantiated and called as such:
 * var obj = WordDictionary()
 * obj.addWord(word)
 * var param_2 = obj.search(word)
 */
